File: src/server/orchestrator/titleWatcher.ts
import fs from 'node:fs';

export interface TitleWatcherOptions {
  path: string;
  pollMs?: number;
  onTitle: (title: string) => void;
  onError: (reason: string) => void;
}

export class TitleWatcher {
  private opts: Required<Omit<TitleWatcherOptions, 'onTitle' | 'onError'>> & TitleWatcherOptions;
  private timer: NodeJS.Timeout | null = null;
  private running = false;
  private lastEmitted: string | null = null;

  constructor(opts: TitleWatcherOptions) {
    this.opts = { pollMs: 1000, ...opts };
  }

  start(): void {
    if (this.running) return;
    this.running = true;
    const tick = () => {
      if (!this.running) return;
      try { this.readOnce(); } catch (e) { this.opts.onError(String(e)); }
      this.timer = setTimeout(tick, this.opts.pollMs);
    };
    this.timer = setTimeout(tick, 0);
  }

  async stop(): Promise<void> {
    this.running = false;
    if (this.timer) { clearTimeout(this.timer); this.timer = null; }
    try { this.readOnce(); } catch (e) { this.opts.onError(String(e)); }
  }

  private readOnce(): void {
    let raw: string;
    try {
      raw = fs.readFileSync(this.opts.path, 'utf8');
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return;
      throw e;  // forwarded to onError by tick-level catch
    }
    const trimmed = raw.trim().slice(0, 80);
    if (trimmed.length === 0) return;
    if (trimmed === this.lastEmitted) return;
    this.lastEmitted = trimmed;
    this.opts.onTitle(trimmed);
  }
}

File: server-elixir/lib/fbi/orchestrator/title_watcher.ex
defmodule FBI.Orchestrator.TitleWatcher do
  @moduledoc "Port of src/server/orchestrator/titleWatcher.ts."
  use GenServer

  defstruct [:path, :poll_ms, :on_title, :last_emitted, :timer]

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid), do: GenServer.stop(pid)

  @impl true
  def init(opts) do
    state = %__MODULE__{
      path: Keyword.fetch!(opts, :path),
      poll_ms: Keyword.get(opts, :poll_ms, 1000),
      on_title: Keyword.fetch!(opts, :on_title),
      last_emitted: nil,
      timer: nil
    }

    {:ok, schedule(state)}
  end

  @impl true
  def handle_info(:tick, state) do
    state = read_once(state)
    {:noreply, schedule(state)}
  end

  @impl true
  def terminate(_reason, state) do
    if state.timer, do: Process.cancel_timer(state.timer)
    read_once(state)
    :ok
  end

  defp schedule(state) do
    timer = Process.send_after(self(), :tick, state.poll_ms)
    %{state | timer: timer}
  end

  defp read_once(%{path: path, on_title: on_title, last_emitted: last} = state) do
    case File.read(path) do
      {:ok, raw} ->
        trimmed = raw |> String.trim() |> String.slice(0, 80)

        if trimmed != "" and trimmed != last do
          on_title.(trimmed)
          %{state | last_emitted: trimmed}
        else
          state
        end

      {:error, :enoent} ->
        state

      _ ->
        state
    end
  end
end
